### PR TITLE
Rescale raster algorithm for Processing

### LINKED ISSUE #26099: Add a Rescale rasters alg
Author Name: **Paolo Cavallini** (@pcav)
Original Redmine Issue: [18208](https://issues.qgis.org/issues/18208)

Redmine category:analysis_library

---
It would be useful to have a new alg rescaling raster values to percent of max.
This would be useful e.g. for probability models, so that we can easily draw lines encompassing a certain percentage of total probability.
See #17761

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Add Rescale raster algorithm to change raster value range preserving shape of the raster's histogram. Useful when rasters from different sources should be compared or processed together and their pixel values should be in the same range. By default algorithm preserves NODATA value of the input raster but there is an option to override it.

Here is histogram of the DEM (value range is 85-243)
histogram-1
And histogram of same DEM rescaled to the range 100-1000
histogram-2

Fixes #26099.

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+QString QgsRescaleRasterAlgorithm::displayName() const
+{
+  return QObject::tr( "Rescale raster" );
+}
+
+QStringList QgsRescaleRasterAlgorithm::tags() const
+{
+  return QObject::tr( "raster,rescale,min,max" ).split( ',' );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmrescaleraster.cpp:37]:
```suggestion
  return QObject::tr( "raster,rescale,minimum,maximum,range" ).split( ',' );
```

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+QString QgsRescaleRasterAlgorithm::groupId() const
+{
+  return QStringLiteral( "rasteranalysis" );
+}
+
+QString QgsRescaleRasterAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Rescales raster layer to the new values range preserving shape "
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmrescaleraster.cpp:52]:
```suggestion
  return QObject::tr( "Rescales raster layer to a new value range, while preserving the shape "
```

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                      "(distribution) of the raster's histogram (pixel values).\n\n"
+                      "By default algorithm preserves original NODATA value, but there is "
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmrescaleraster.cpp:54]:
```suggestion
                      "By default the algorithm preserves original the NODATA value, but there is "
```

**Review by nyalldawson — APPROVED** [review]

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+QString QgsRescaleRasterAlgorithm::groupId() const
+{
+  return QStringLiteral( "rasteranalysis" );
+}
+
+QString QgsRescaleRasterAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Rescales raster layer to the new values range preserving shape "
+                      "(distribution) of the raster's histogram (pixel values).\n\n"
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmrescaleraster.cpp:53]:
```suggestion
                      "(distribution) of the raster's histogram (pixel values) Input values are mapped using a linear interpolation from the source raster's minimum and maximum pixel values to the destination minimum and maximum pixel range.\n\n"
```

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp (hunk 1 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmrescaleraster.cpp]: Fixed

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp (hunk 3 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmrescaleraster.cpp]: Fixed, thanks!

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp (hunk 4 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmrescaleraster.cpp]: Applied, thanks!

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp (hunk 2 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmrescaleraster.cpp]: Fixed.

**Comment by Saijin-Naib** [comment]
This has some awesome Remote Sensing applications!

**Pull request merged**